modules/apps/static/portal-osgi-web/portal-osgi-web-servlet-jsp-compiler/src/main/java/com/liferay/portal/osgi/web/servlet/jsp/compiler/internal/JspTaglibIDInitializer.java

```diff
@@ -0,0 +1,106 @@
+/**
```

Comment:
Hi @dantewang,
Is it possible to include these files directly into the bundle by using  -includeresource in bnd.bnd like this one:`-includeresource: @javax.servlet.jsp.jstl-[0-9.]*.jar!/META-INF/*.tld`?

Tina.

Reply:
@tinatian 
I can't remember all the details at the moment, but I think this commit is to resolve Tomcat Jasper's issues on other app servers, like JBoss/Wildfly.

The problem here is that Tomcat's `DigesterFactory` tries to find locations of the tlds/xsds using `ServletContext`'s class loader, and then `JspContext`'s class loader. This logic is actually tightly coupled with Tomcat. A standard servlet-api jar or jsp-api jar won't have some of the tld/xsd files, but Tomcat's servlet-api.jar has. Thus the finding logic works well on Tomcat but fails on other app servers.

Reply:
@tinatian And we have already included the files because they are exactly included by that `-includeresource`. This class just helps Tomcat's `DigesterFactory` find these files.

Reply:
@dantewang
Can we avoid using Tomcat's TldParser?

Tina.

Reply:
@tinatian I didn't try that. I'll look into it.

Reply:
@tinatian I took a look at Tomcat Jasper. The problem here is not related to `TldParser`. If `DigesterFactory` doesn't have the information, JSP parsing will fail -- even if the TLD cache has all the TLDs we provide. This logic is in `JspDocumentParser` which is a package-private class tightly coupled in the process of translating JSP file to Java file.